Objects/genobject.c

```diff
@@ -1161,7 +1163,47 @@ PyTypeObject _PyCoroWrapper_Type = {
 PyObject *
 PyCoro_New(PyFrameObject *f, PyObject *name, PyObject *qualname)
 {
-    return gen_new_with_qualname(&PyCoro_Type, f, name, qualname);
+    PyObject *coro = gen_new_with_qualname(&PyCoro_Type, f, name, qualname);
+    if (!coro)
+        return NULL;
+
+    PyThreadState *tstate = PyThreadState_GET();
+    int depth = tstate->coroutine_origin_tracking_depth;
+
+    if (depth == 0) {
+        ((PyCoroObject *)coro)->cr_origin = NULL;
+    } else {
+        PyObject *origin = PyList_New(depth);
+        /* Immediately pass ownership to coro, so on error paths we don't have
+           to worry about it separately. */
+        ((PyCoroObject *)coro)->cr_origin = origin;
```

Comment:
I just realized that exposing list object directly through `cr_origin` means that users can mutate it.  How about we cycle through the frames two times: 1st time to determine the real depth we can capture; 2nd time to populate a `cr_origin` **tuple** (not list). Traversing a relatively short chain of frames should be pretty fast, and this is a debug mode after all. What do you think?

Reply:
Another option would be to keep things as is, but to change `cr_origin` to a getter, and return a copy of the list every time. But I like the tuple idea more.

Reply:
Right, I originally left out the `Py_VISIT` call and then realized that mutation was possible, so figured I'd put it in just in case :-).

I don't think it matters too much whether we expose a tuple, copy the list, or just keep the current thing where someone can mutate it if they really want to. (The warnings code is robust against such shenanigans: worst case it'll print an error b/c `cr_origin` was corrupt, and then the real warning without a traceback.) Given this I have a mild preference not to do the tuple thing, because it leads to the most complicated C code. Doing an extra loop is totally doable of course, but it's also plenty complicated to hide some kind of off-by-one bug or something.